# Question 1

In [1]:
import numpy as np # importing numpy library
import pandas as pd # importing pandas library
import warnings # importing warnings
warnings.filterwarnings("ignore")

In [2]:
train = pd.read_csv('train.csv') # reading train dataset and storing it as train
test = pd.read_csv('test.csv') # reading test dataset and storing it as test

In [3]:
train['Sex'] = train['Sex'].replace(["female", "male"], [1, 0]) # replacing female and male as 1 and 0 respectively
train['Embarked'] = train['Embarked'].replace(['S','C','Q'],[0,1,2]) # replacing S,C,Q with 0,1,2 respectively

In [4]:
train.fillna(train.mean(),axis=0,inplace=True) # filling all the NaN values with mean of respective columns

### 1. Find the correlation between ‘survived’ (target column) and ‘sex’ column for the Titanic use case in class

In [5]:
train['Survived'].corr(train['Sex']) # finding the correlation between 'Survived' column and 'Sex' column

0.5433513806577546

### a. Do you think we should keep this feature?
###### Ans:  Yes, I think we need to keep this feature. As we replaced the values of Sex column from female and male to 1 and 0 respectively, we can include this column while giving the input to the model

### 2. Do at least two visualizations to describe or show correlations.
###### Showing correlations

In [6]:
train.corr().style.background_gradient(cmap="Greens") # finding the correlation for train dataframe

### 3. Implement Naïve Bayes method using scikit-learn library and report the accuracy.

In [7]:
X = train.drop(columns = ['Name','Survived','Ticket','Cabin']) # considering only the columns which are of numeric type
Y = train[['Survived']] # considering the target column

In [8]:
from sklearn.metrics import confusion_matrix # importing confusion_matrix function
from sklearn.metrics import classification_report # importing classification_report function
from sklearn.model_selection import train_test_split # importing train_test_split function

In [9]:
# spliting the given dataset into train and test datasets
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2, random_state = 0)

In [10]:
# Gaussian Naive Bayes
from sklearn.naive_bayes import GaussianNB # importing GaussianNB 
classifier = GaussianNB() # creating the model
classifier.fit(X_train, Y_train) # feeding model with training dataset
Y_pred = classifier.predict(X_test) # predicting the dependent variable in the test dataset
# Summary of the predictions made by the classifier
print(classification_report(Y_test, Y_pred)) # printing the classification report
print(confusion_matrix(Y_test, Y_pred)) # printing the confusion matrix
# Accuracy score
from sklearn.metrics import accuracy_score # importing the accuracy_score
print('accuracy is',accuracy_score(Y_pred,Y_test)) # printing the accuracy of predicted values with true values

              precision    recall  f1-score   support

           0       0.84      0.82      0.83       110
           1       0.72      0.75      0.74        69

    accuracy                           0.79       179
   macro avg       0.78      0.79      0.78       179
weighted avg       0.80      0.79      0.79       179

[[90 20]
 [17 52]]
accuracy is 0.7932960893854749


# Question 2

### 1. Implement Naïve Bayes method using scikit-learn library.

### a. Use the glass dataset available in Link also provided in your assignment

In [11]:
data = pd.read_csv('glass.csv') # importing glass dataset and storing it as data 

In [12]:
X = data.drop(columns=['Type']) # considering the only the independent columns
Y = data['Type'] # considering the dependent column

### b. Use train_test_split to create training and testing part.

In [13]:
# spliting the given dataset into train and test datasets
X_train, X_test, Y_train, Y_test = train_test_split(X, Y, test_size = 0.2, random_state = 0)

### 2. Evaluate the model on testing part using score and classification_report

In [14]:
# Gaussian Naive Bayes
from sklearn.naive_bayes import GaussianNB # importing GaussianNB 
classifier = GaussianNB() # creating the model
classifier.fit(X_train, Y_train) # feeding model with training dataset
Y_pred = classifier.predict(X_test) # predicting the dependent variable in the test dataset
# Summary of the predictions made by the classifiera
print(classification_report(Y_test, Y_pred)) # printing the classification report
print(confusion_matrix(Y_test, Y_pred)) # printing the confusion matrix
# Accuracy score
from sklearn.metrics import accuracy_score # importing the accuracy_score
print('accuracy is',accuracy_score(Y_pred,Y_test)) # printing the accuracy of predicted values with true values

              precision    recall  f1-score   support

           1       0.19      0.44      0.27         9
           2       0.33      0.16      0.21        19
           3       0.33      0.20      0.25         5
           5       0.00      0.00      0.00         2
           6       0.67      1.00      0.80         2
           7       1.00      1.00      1.00         6

    accuracy                           0.37        43
   macro avg       0.42      0.47      0.42        43
weighted avg       0.40      0.37      0.36        43

[[ 4  3  1  0  1  0]
 [14  3  1  1  0  0]
 [ 3  1  1  0  0  0]
 [ 0  2  0  0  0  0]
 [ 0  0  0  0  2  0]
 [ 0  0  0  0  0  6]]
accuracy is 0.37209302325581395


### 1. Implement linear SVM method using scikit library

In [15]:
# Support Vector Machine's 
from sklearn.svm import SVC # importing SVC 
classifier = SVC() # creating the model
classifier.fit(X_train, Y_train) # feeding model with training dataset
Y_pred = classifier.predict(X_test) # predicting the dependent variable in the test dataset

### 2. Evaluate the model on testing part using score and classification_report

In [16]:
# Summary of the predictions made by the classifier
print(classification_report(Y_test, Y_pred)) # printing the classification report
print(confusion_matrix(Y_test, Y_pred)) # printing the confusion matrix
# Accuracy score
from sklearn.metrics import accuracy_score # importing the accuracy_score
print('accuracy is',accuracy_score(Y_pred,Y_test)) # printing the accuracy of predicted values with true values

              precision    recall  f1-score   support

           1       0.21      1.00      0.35         9
           2       0.00      0.00      0.00        19
           3       0.00      0.00      0.00         5
           5       0.00      0.00      0.00         2
           6       0.00      0.00      0.00         2
           7       0.00      0.00      0.00         6

    accuracy                           0.21        43
   macro avg       0.03      0.17      0.06        43
weighted avg       0.04      0.21      0.07        43

[[ 9  0  0  0  0  0]
 [19  0  0  0  0  0]
 [ 5  0  0  0  0  0]
 [ 2  0  0  0  0  0]
 [ 2  0  0  0  0  0]
 [ 6  0  0  0  0  0]]
accuracy is 0.20930232558139536


### Do at least two visualizations to describe or show correlations in the Glass Dataset

In [17]:
data.corr().style.background_gradient(cmap="Greens")

### Which algorithm you got better accuracy? Can you justify why?
###### Ans: From the above result, it is observed that, the accuracy from Naive Bayes algorithm (37.2%) is greater when compared with the accuracy from SVM algorithm(20.9%) So Navie Bayers is better than SVM